index.js

```diff
@@ -104,51 +77,279 @@ module.exports = {
+   *
+   * @typedef configuration
+   * @type {object}
+   * @property {?string[]} httpAddresses
+   * @property {?string[]} wsAddresses
+   * @property {?string[]} ipcAddresses
+   * @property {?number} connectionTimeout
+   * @property {?function(Error):void} errorHandler - called when an otherwise unhandled asynchronous error occurs during the course of operation.
+   *
+   * @param {!configuration} configuration
+   * @param {!function(?Error):void} initialConnectCallback - if the error parameter is null then the connection was successful
+   * @returns {void}
+   */
+  connect: function (configuration, initialConnectCallback) {
+    this.resetState();
+
+    // overwrite configuration values with user config, throw away unused user config
+    for (var key in this.configuration) {
+      if (configuration[key] === undefined || configuration[key] === null) continue;
+      this.configuration[key] = configuration[key];
+    }
+
+    // validate configuration
+    if (typeof this.configuration.errorHandler !== "function") throw new Error("configuration.errorHandler must be a function");
+    if (!Array.isArray(this.configuration.httpAddresses)) return this.configuration.errorHandler(new Error("configuration.httpAddresses must be an array."));
+    if (this.configuration.httpAddresses.some(function (x) { return typeof x !== "string"; })) return this.configuration.errorHandler(new Error("configuration.httpAddresses must contain only strings."));
+    if (!Array.isArray(this.configuration.wsAddresses)) return this.configuration.errorHandler(new Error("configuration.wsAddresses must be an array."));
+    if (this.configuration.wsAddresses.some(function (x) { return typeof x !== "string"; })) return this.configuration.errorHandler(new Error("configuration.wsAddresses must contain only strings."));
+    if (!Array.isArray(this.configuration.ipcAddresses)) return this.configuration.errorHandler(new Error("configuration.ipcAddresses must be an array."));
+    if (this.configuration.ipcAddresses.some(function (x) { return typeof x !== "string"; })) return this.configuration.errorHandler(new Error("configuration.ipcAddresses must contain only strings."));
+
+    var syncOnly = !initialConnectCallback;
+    if (syncOnly) initialConnectCallback = function (error) { if (error instanceof Error) throw error; else if (error) throw new ErrorWithData(error); };
+
+    // initialize the transporter, this will be how we send to and receive from the blockchain
+    new Transporter(this.configuration, this.internalState.shimMessageHandler, syncOnly, function (error, transporter) { // jshint ignore:line
+      if (error !== null) return initialConnectCallback(error);
+      this.internalState.transporter = transporter;
+      // ensure we can do basic JSON-RPC over this connection
+      this.version(function (errorOrResult) {
+        if (errorOrResult instanceof Error || errorOrResult.error) return initialConnectCallback(errorOrResult);
```

Comment:
Unclear.  I make it a policy of always wrapping errors in `Error` objects, so everywhere in the code I have written, if there is an error it is of type `Error`.  However, I realized (after implementing this originally) that the rest of the code does not follow this pattern everywhere and will instead sometimes return an object that has an `error` property on it.  Rather than trying to figure out all of the places that can return one way or the other, I decided to have all of the checks for errors be like this so they are robust in either case.

It should never be undefined, if it is then there is a programming error.  Null is a more interesting question though, I don't believe it is correct for `version` to return a null result, though I suppose the protocol probably doesn't explicitly state otherwise?  I think the result parser of yours will turn a `null` into an object with an `error` on it though.

If you want, we can add null checking here so this code is resilient to the parser being changed to allow nulls through.  At the moment I don't believe we are at risk of getting a null from this function.